Port to glutin 0.21

est31 opened the pull request:
Trying to port to the upcoming glutin release. @ZeGentzy I'm not sure how to fix the `make_current` related compilation errors. Any ideas?

ZeGentzy commented:
I'd rather glium changes their `make_current` to also take `self`, however, I understand if you don't want to break downstream users' code.

Anyways, here is a patch to apply to your commit to make it compile: https://termbin.com/24z4a

Please consider also exposing `make_not_current`, and possibly implementing `Send + Sync`.

est31 commented:
Rebased, but have to apply the suggestions by @ZeGentzy first so it's still WIP.

est31 commented:
@ZeGentzy that link is dead now, could you re-create it?

ZeGentzy commented:
Sorry. Don't have it.

est31 commented:
@ZeGentzy that blog post is very helpful thank you! I think I can figure it out from here on.

est31 commented:
@mitchmindtree it's ready for review now

mitchmindtree commented:
It LGTM though I don't fully understand the glutin changes - if @ZeGentzy can give the all clear I'm happy to merge :+1: *(I'll merge anyway if it takes too long)*.

ZeGentzy commented on src/backend/glutin/mod.rs:
```diff
@@ -127,21 +130,24 @@ impl Display {
         &self,
         wb: glutin::WindowBuilder,
-        cb: glutin::ContextBuilder,
+        cb: glutin::ContextBuilder<NotCurrent>,
```
Should be T, not `NotCurrent`. What if they want to pass a cb which shares with a current context?

ZeGentzy commented on src/backend/glutin/mod.rs:
```diff
@@ -127,21 +130,24 @@ impl Display {
+            cb.build_windowed(wb, events_loop)?
+        };
+        let new_gl_window = unsafe {
+            new_gl_window.treat_as_current()
```
This context is not current. You are treating it as current, even though it is not. You are then placing it into the takeable.

This is not incorrect, per se, but it also fails to utilize glutin's context currency tracking system.

est31 replied:
@ZeGentzy yeah this port is definitely not perfect. Probably in order to do it properly, one should store a `glutin::WindowedContext<T>` everywhere but this would most likely be very involved especially for users as suddenly the glium context is taken by move and stuff.

ZeGentzy commented on src/backend/glutin/mod.rs:
```diff
@@ -63,53 +63,56 @@ impl Display {
 
     fn new_inner(
-        gl_window: glutin::WindowedContext,
+        gl_window: glutin::WindowedContext<Pc>,
```
Context is assumed to be current. Fine and all, but what if they want to pass a not current one. This forces users to either pretend it is current with `treat_as_current` or actually make it current, with `make_current`, which might not be something they want.

This is not wrong per se, indeed this code would be functional.

est31 commented:
@ZeGentzy I've addressed your comments by adding generic params for the constructor functions. Now users should have more flexibility at what they pass to the glium data structures.

ZeGentzy commented on src/backend/glutin/headless.rs:
```diff
@@ -72,41 +77,41 @@ impl Headless {
     /// Performs a compatibility check to make sure that all core elements of glium are supported
     /// by the implementation.
-    pub fn new(context: glutin::Context) -> Result<Self, IncompatibleOpenGl> {
+    pub fn new(context: glutin::Context<Pc>) -> Result<Self, IncompatibleOpenGl> {
```
Should not all the `Pc`s in this file also be `T`s?

est31 commented:
I made all the headless constructors generic now.

est31 commented:
@ZeGentzy is it fine now?

ZeGentzy commented on src/backend/glutin/mod.rs:
```diff
@@ -61,55 +61,58 @@ impl Display {
     /// The same as the `unchecked` constructor, but allows for specifying debug callback behaviour.
     pub unsafe fn unchecked_with_debug(
-        gl_window: glutin::WindowedContext,
+        gl_window: glutin::WindowedContext<Pc>,
```
Here.

ZeGentzy commented on src/backend/glutin/mod.rs:
```diff
@@ -61,55 +61,58 @@ impl Display {
     /// This function does the same as `build_glium`, except that the resulting context
     /// will assume that the current OpenGL context will never change.
-    pub unsafe fn unchecked(gl_window: glutin::WindowedContext) -> Result<Self, IncompatibleOpenGl> {
+    pub unsafe fn unchecked(gl_window: glutin::WindowedContext<Pc>) -> Result<Self, IncompatibleOpenGl> {
```
Here?

mitchmindtree commented:
Thanks folks!

est31 commented:
@mitchmindtree @ZeGentzy thanks for the review and help!
